### ISSUE TITLE
Visualizer anti-aliasing

### ISSUE THREAD

**Issue by chrisdembia** [simbody/simbody#381]
Antialiasing can be achieved with 2 lines

```
glutInitDisplay(... | GLUT_MULTISAMPLE);
glEnable(GL_MULTISAMPLE);
```

Here's before:

![screenshot from 2015-06-03 13 05 25](https://cloud.githubusercontent.com/assets/846001/7970258/98a15656-09f1-11e5-94ec-e8f0a9092a67.png)

and after:
![screenshot from 2015-06-03 13 05 18](https://cloud.githubusercontent.com/assets/846001/7970257/98a03064-09f1-11e5-9008-f68fee246283.png)

Is this worthwhile?

**Comment by sherm1** [comment]
Sweet! Yes, that would be great especially for making nice stills.

It should be optional though, I think, because it is likely to have a large performance impact on complicated scenes. @peastman says it operates by generating multiple renderings of the same scene and combining them. It would be easy to add an antialiasing option to the Visualizer -- could even add it to the "View" button that is always present so it wouldn't have to be done programmatically.

**Comment by peastman** [comment]
It's not exactly that it generates multiple renderings of the same scene.  It's more sophisticated than that.  But there is a performance cost.  Given the sorts of scenes you're rendering, though, that's likely to be a minor issue.  Modern GPUs are designed to handle *somewhat* more complicated scenes than the one shown above. :)

**Comment by chrisdembia** [comment]
It wasn't obvious to me how to get Visualizer settings into simbody-visualizer.cpp:main, which is where I tried adding the new lines. I think having the option in the window is good but I think a programmatic option would be good as well.

**Comment by sherm1** [comment]
> It wasn't obvious to me how to get Visualizer settings into simbody-visualizer.cpp:main

Maybe that's not necessary. What if in main you put:
```c++
glutInitDisplayMode(... | GLUT_MULTISAMPLE);
glDisable(GL_MULTISAMPLE);
shouldAntialias = false;
```
(I'm assuming that disabling multisample avoids the performance cost -- do you know, Peter?)

Then add an "Antialias" entry to the View menu, a toggle similar to the "Show shadows" entry (see MENU_SHOW_SHADOWS in the visualizer code). If a user selects that you could execute:
```c++
shouldAntialias = !shouldAntialias;
if (shouldAntialias) glEnable(GL_MULTISAMPLE);
else glDisable(GL_MULTISAMPLE);
```

If that works, programmatic control could be provided also, but it would be separate. You would add a new message to the many already supported by the VisualizerProtocol and have it execute the same code that the View menu item would do.

**Comment by peastman** [comment]
I think that should be fine.  By the way, you can find a discussion of this antialiasing method at https://www.opengl.org/wiki/Multisampling.

**Comment by chrisdembia** [comment]
Thanks both of you. @sherm1, if I test this on Linux and OSX can you test on Windows?

**Comment by sherm1** [comment]
> By the way, you can find a discussion of this antialiasing method at https://www.opengl.org/wiki/Multisampling.

I noticed in there that they use `glEnable(GL_MULTISAMPLE_ARB)` rather than just `GL_MULTISAMPLE`. Not sure what that means. Also it would be good to check when multisample was added to OpenGL to make sure we don't accidentally prevent the visualizer from working if someone has an older OpenGL.

**Comment by sherm1** [comment]
> if I test this on Linux and OSX can you test on Windows

Yes. I only have OpenGL 4.4 so can't check anything earlier. I *think* multisample was added in OpenGL 1.5 so we are probably safe for everything but raw Microsoft opengl with which the visualizer doesn't work anyway.

**Comment by peastman** [comment]
According to https://stackoverflow.com/questions/4207506/where-is-gl-multisample-defined, it became a standard feature in 1.3.  Extensions start out being proposed by one vendor, and they have their own name for it.  Then if everyone agrees it's useful, the OpenGL Architecture Review Board standardizes it and gives it a name with the _ARB suffix.  This is sort of an intermediate state.  It's still an extension, so implementations don't have to include it.  But if they do, they're supposed to use the standard interface.  And then if it later becomes a core feature, they define a new name for it without the _ARB extension, but also keep supporting the old name.

**Comment by sherm1** [comment]
Ah, thanks Peter. So no "_ARB" for us.

Chris, I would suggest just doing the View menu option since that is entirely local to the visualizer. That would be all that's needed for making nice pictures. You could leave the programmatic access until someone has a need for it.

**Comment by chrisdembia** [comment]
This documentation might help: This page http://docs.gl/ might
